### PR TITLE
Add ecdsa

### PR DESCRIPTION
**Pull request by elliemdaw**
- This hasn't been fully tested with the NIST server since I haven't been able to successfully retrieve a vector set
- Adds the library side for ECDSA
- Has a stub for the ECDSA handler in the app code

### CODE REVIEW

File: app/app_main.c [hunk 1]
```diff
@@ -1194,9 +1201,76 @@ static void enable_rsa (ACVP_CTX *ctx) {
...
+    rv = acvp_enable_ecdsa_cap_parm(ctx, ACVP_ECDSA_SIGVER, ACVP_CURVE, "k-163");
+    CHECK_ENABLE_CAP_RV(rv);
+    rv = acvp_enable_ecdsa_cap_parm(ctx, ACVP_ECDSA_SIGVER, ACVP_CURVE, "b-163");
+    CHECK_ENABLE_CAP_RV(rv);
+    rv = acvp_enable_ecdsa_cap_parm(ctx, ACVP_ECDSA_SIGVER, ACVP_HASH_ALG, "SHA2-224");
+    CHECK_ENABLE_CAP_RV(rv);
 #endif
+}
 
-#ifdef ACVP_NO_RUNTIME
```
Comment (bfussell) [inline app/app_main.c]: Is DRBG allowed at runtime now ? Or maybe blocked somewhere else ?

File: src/acvp_ecdsa.c [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        strncpy((char *)stc->s, s, strnlen(s, 128));
+        strncpy((char *)stc->r, r, strnlen(r, 128));
+    }
+    if (cipher == ACVP_ECDSA_SIGVER || cipher == ACVP_ECDSA_SIGGEN) {
+        stc->message = calloc(ACVP_RSA_EXP_LEN_MAX, sizeof(char));
+        if (!stc->message) { return ACVP_MALLOC_FAIL; }
+        strncpy((char *)stc->message, message, strnlen(message, ACVP_RSA_MSGLEN_MAX));
+    }
+    stc->d = calloc(ACVP_RSA_EXP_LEN_MAX, sizeof(char));
+    if (!stc->d) { return ACVP_MALLOC_FAIL; }
```
Comment (bfussell) [inline src/acvp_ecdsa.c]: Do we need to spit out an error message on the failure so we know what type it is or does the caller do that ?

File: src/acvp_ecdsa.c [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+             */
+            r_tval = json_value_init_object();
+            r_tobj = json_value_get_object(r_tval);
+            
+            json_object_set_number(r_tobj, "tcId", tc_id);
+            
+            rv = acvp_ecdsa_init_tc(ctx, alg_id, &stc, tc_id, curve, secret_gen_mode, qx, qy, message, r, s);
+            
+            /* Process the current test vector... */
+            if (rv == ACVP_SUCCESS) {
```
Comment (bfussell) [inline src/acvp_ecdsa.c]: On failure should we exit out here ?

File: src/acvp_ecdsa.c [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    json_result = json_serialize_to_string_pretty(ctx->kat_resp);
+    if (ctx->debug == ACVP_LOG_LVL_VERBOSE) {
+        printf("\n\n%s\n\n", json_result);
+    } else {
+        ACVP_LOG_INFO("\n\n%s\n\n", json_result);
+    }
+    json_free_serialized_string(json_result);
+    
+    return rv;
+}
```
Comment (bfussell) [inline src/acvp_ecdsa.c]: no newline at end of file

File: src/acvp_ecdsa.c [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            }
+            
+            /*
+             * Release all the memory associated with the test case
+             */
+            key_err:
+            acvp_ecdsa_release_tc(&stc);
+            
+            /* Append the test response value to array */
+            json_array_append_value(r_tarr, r_tval);
```
Comment (bfussell) [inline src/acvp_ecdsa.c]: Should we try and append on a failure ?

File: app/app_main.c (hunk 1 above)
Comment (elliemdaw) [reply app/app_main.c]: Oh that looks confusing in the diff... I made them all a part of the same #ifdef so DRBG, ECDSA, and RSA are all in the single block.

File: src/acvp_ecdsa.c (hunk 2 above)
Comment (elliemdaw) [reply src/acvp_ecdsa.c]: I will add more verbose logs here :)

**Comment by elliemdaw** [comment]
@bfussell I pushed up changes from these comments... the logging in acvp_ecdsa to initialize a test case is generic unless the calloc fails for the sig specific buffer

**Pull request merged**